src/decompose.rs

```diff
@@ -93,21 +103,28 @@ impl<I: Iterator<Item=char>> Iterator for Decompositions<I> {
 
     #[inline]
     fn next(&mut self) -> Option<char> {
-        while self.ready == 0 && !self.done {
+        while self.ready.end == 0 {
             match (self.iter.next(), &self.kind) {
                 (Some(ch), &DecompositionType::Canonical) => {
                     super::char::decompose_canonical(ch, |d| self.push_back(d));
-                },
+                }
                 (Some(ch), &DecompositionType::Compatible) => {
                     super::char::decompose_compatible(ch, |d| self.push_back(d));
-                },
+                }
                 (None, _) => {
-                    self.sort_pending();
-                    self.done = true;
-                },
+                    if self.buffer.is_empty() {
+                        return None;
+                    } else {
+                        self.sort_pending();
+                        break;
+                    }
+                }
             }
         }
-        self.pop_front()
+
+        let (_, ch) = self.buffer[self.ready.start];
```

Comment:
How do you feel about solving this with a comment instead of a restructuring?

- The `pop_front` idea is elegant but benchmarks worse. I believe this is because we add an extra branch to the beginning of every `next` call in the common case where we are buffering only one character at a time so that we need to advance the underlying iterator each time.

- Even changing `self.ready.end == 0` to `self.ready.end == self.ready.start` benchmarks slightly worse on my machine; I'm not sure why and am not much good at reading assembly, but perhaps there's an extra load, or the extra efficiency of a comparison to zero is meaningful.

Reply:
Yeah, a comment sounds good!  In addition to a correctness argument, can you include a note explaining that it's written this way for performance?  That way, subsequent editors know to run benchmarks to check to see if there's a regression.

Thanks for giving it a shot @djudd :)